library/alloc/src/string.rs

```diff
@@ -1354,9 +1354,14 @@ impl String {
     #[inline]
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn push(&mut self, ch: char) {
-        match ch.len_utf8() {
-            1 => self.vec.push(ch as u8),
-            _ => self.vec.extend_from_slice(ch.encode_utf8(&mut [0; 4]).as_bytes()),
+        let len = self.len();
+        let ch_len = ch.len_utf8();
+        self.reserve(ch_len);
```

Comment:
Reserving 4 bytes and doing a single store [makes](https://rust.godbolt.org/z/3ooY7xcnW) little difference other than getting rid of the unsafe, but reserving 4 bytes and doing the same writes [makes](https://rust.godbolt.org/z/13ov7PW1P) the non-reallocating path 20% instructions shorter. However, it may cause the string to take up extra space (say, a 3-byte char is pushed to a 61-byte string). Is this acceptable?

Reply:
Worth noting here that this effectively just means that you'll have up to three extra bytes reserved always, since reserve takes into account existing capacity. It does slow down the case of, say, adding a newline to an existing string, but it would speed up repeated insertions, which are probably the bigger performance hit.

Reply:
This is an option, but I don't want to change the behavior of `String::push` in this PR. Performance doesn't seem to be affected much.